File: lectures/Clase 03 (07-09-2018) - Verificación de programas secuenciales/progs_clase_03/divmod_v8.c
#include <assert.h>
#include <stdio.h>

void divmod(int x, int y, int* q, int* r)
{
    assert(0<=x && 0<y);
    *r=x; *q=0;
    assert(0<=*r && 0<y && x==y*(*q)+(*r));
    while (*r>=y) {
        assert(0<=*r && 0<y && y<=*r && x==y*(*q)+(*r));
        *r=*r-y; (*q)++;
        assert(0<=*r && 0<y && x==y*(*q)+(*r));
    }
    assert(0<=*r && *r<y && x==y*(*q)+(*r));
}

void main(void)
{
    int x, y=7, q, r, i;

    for (i=0; i<7; i++) {
        x=100+i;
        divmod(x,y,&q,&r); 
        printf("%d = %d*%d + %d\n",x,y,q,r);
    }
}

File: lectures/Clase 03 (07-09-2018) - Verificación de programas secuenciales/progs_clase_03/divmod_v8.py
def divMod(x,y):
    assert x>=0 and y>0, "Wrong input"  # precondition
    r,q= x,0
    while r>=y:
        assert r>=0 and y>0 and y<=r and x==y*q+r, "Invariant error"
        r,q= r-y,q+1
        assert r>=0 and y>0 and x==y*q+r, "Invariant error"
    assert x==y*q+r and 0<=r and r<y, "Function error"   # postcondition
    return q,r

for i in range(7):
    q,r= divMod(100+i,7)
    print("%d = %d*%d + %d" % (100+i,7,q,r))
